extensions/arc/src/ui/dialogs/connectControllerDialog.ts

```diff
@@ -5,37 +5,73 @@
 
 import { ControllerInfo } from 'arc';
 import * as azdata from 'azdata';
+import * as azdataExt from 'azdata-ext';
 import { v4 as uuid } from 'uuid';
 import * as vscode from 'vscode';
 import { Deferred } from '../../common/promise';
 import * as loc from '../../localizedConstants';
 import { ControllerModel } from '../../models/controllerModel';
 import { InitializingComponent } from '../components/initializingComponent';
 import { AzureArcTreeDataProvider } from '../tree/azureArcTreeDataProvider';
+import { getErrorMessage } from '../../common/utils';
 
 export type ConnectToControllerDialogModel = { controllerModel: ControllerModel, password: string };
 export class ConnectToControllerDialog extends InitializingComponent {
 	private modelBuilder!: azdata.ModelBuilder;
+	protected dialog: azdata.window.Dialog;
 
-	private urlInputBox!: azdata.InputBoxComponent;
-	private nameInputBox!: azdata.InputBoxComponent;
-	private usernameInputBox!: azdata.InputBoxComponent;
-	private passwordInputBox!: azdata.InputBoxComponent;
-	private rememberPwCheckBox!: azdata.CheckBoxComponent;
+	protected urlInputBox!: azdata.InputBoxComponent;
+	protected nameInputBox!: azdata.InputBoxComponent;
+	protected usernameInputBox!: azdata.InputBoxComponent;
+	protected passwordInputBox!: azdata.InputBoxComponent;
+	protected rememberPwCheckBox!: azdata.CheckBoxComponent;
 
-	private _completionPromise = new Deferred<ConnectToControllerDialogModel | undefined>();
+	protected _completionPromise = new Deferred<ConnectToControllerDialogModel | undefined>();
+	protected _id!: string;
 
-	private _id!: string;
 
-	constructor(private _treeDataProvider: AzureArcTreeDataProvider) {
+	constructor(protected _treeDataProvider: AzureArcTreeDataProvider, title: string = loc.connectToController) {
 		super();
+		this.dialog = azdata.window.createModelViewDialog(title);
+	}
+
+	private getComponents(isPwReacquisition: boolean): (azdata.FormComponent<azdata.Component> & { layout?: azdata.FormItemLayout | undefined; })[] {
+		const components: (azdata.FormComponent<azdata.Component> & { layout?: azdata.FormItemLayout | undefined; })[] = [
+			{
+				component: this.urlInputBox,
+				title: loc.controllerUrl,
+				required: true
+			}, {
+				component: this.nameInputBox,
+				title: loc.controllerName,
+				required: false
+			}, {
+				component: this.usernameInputBox,
+				title: loc.username,
+				required: true
+			}, {
+				component: this.passwordInputBox,
+				title: loc.password,
+				required: true
+			}
+		];
+		if (!isPwReacquisition) {
+			components.push({
+				component: this.rememberPwCheckBox,
+				title: ''
+			});
+		}
+		return components;
 	}
 
 	public showDialog(controllerInfo?: ControllerInfo, password?: string): azdata.window.Dialog {
+		return this._showDialog(controllerInfo, password);
+	}
+
+	protected _showDialog(controllerInfo: ControllerInfo | undefined, password: string | undefined, isPwReacquisition: boolean = false): azdata.window.Dialog {
```

Comment:
_ should only be used for private members. The pattern vs code uses is something like `doShowDialog` for protected functions which seems generic enough

Reply:
removed the method altogether